framework/src/executioners/Steady.C

```diff
@@ -126,3 +127,15 @@ Steady::checkIntegrity()
   if (_problem.getNonlinearSystemBase().containsTimeKernel())
     mooseError("You have specified time kernels in your steady state simulation");
 }
+
+unsigned int
+Steady::getIterationNumberOutput() const
+{
+  return _optimization_iteration_number;
+}
+
+void
+Steady::setIterationNumberOutput(unsigned int iteration_number)
```

Comment:
This two functions look to me belonging to output system, either in Outputs or OutputWarehouse.

Reply:
If I understand well, the executioner would update an output system member variable that can be used without having the output derived objects pulled the executioner. That makes sense to me. I'd rather not make this change now in this PR. I suggest we leave that for when we further exercise this "special" output capability.

Reply:
This interface function seems the hot spot from reviewers ;-) This really looks like a short term solution. If we can have a little bit more visionary design here, it would be better. I do not have super strong opinions on this.